app/src/main/java/com/example/cocktails/presentation/fragments/details/viewmodel/CocktailDetailsViewModel.kt

```diff
@@ -0,0 +1,151 @@
+                    cocktailDetailsMap[Constants.instructionsKey] ?: "",
+                    cocktailDetailsMap[Constants.drinkThumbKey] ?: ""
+                )
+                _cocktailsDetails.value = Resource.Success(cocktailDetailsData)
+
+            } else {
+                _cocktailsDetails.value =
+                    Resource.Error(application.applicationContext.resources.getString(R.string.failed_fetch))
+            }
+        }
+    }
+
+    fun insertCocktail(cocktail: Cocktail) {
+        viewModelScope.launch(handler) {
+            val response = cocktailRepository.insertCocktail(cocktail)
+            if (response > 0) {
+                cocktailState.value = CocktailState.SuccessAddOrDelete
+            } else {
+                cocktailState.value =
+                    CocktailState.Error(application.applicationContext.resources.getString(R.string.failed_fetch))
+            }
+        }
+    }
+
+    fun deleteCocktail(cocktail: Cocktail) {
+        viewModelScope.launch(handler) {
+            val response = cocktailRepository.deleteCocktail(cocktail)
+            if (response > 0) {
+                cocktailState.value = CocktailState.SuccessAddOrDelete
+            } else {
+                cocktailState.value =
+                    CocktailState.Error(application.applicationContext.resources.getString(R.string.failed_fetch))
+            }
+        }
+    }
+
+    private fun getNonNullFields(cocktailDetailsList: List<CocktailDetails>): Map<String, String> {
+        return cocktailDetailsList.firstOrNull()?.let { cocktailDetails ->
+            val fields = CocktailDetails::class.java.declaredFields
+            val result = mutableMapOf<String, String>()
+
+            for (field in fields) {
+                field.isAccessible = true
+                val value = field.get(cocktailDetails)
+                if (value != null) {
+                    result[field.name] = value.toString()
+                }
+            }
+
+            result.ifEmpty { null }
+        } ?: mapOf()
+    }
+
+    fun createCocktail(
+        cocktailDetailsData: CocktailDetailsData,
+        id: Int,
+        argFavorite: Boolean,
+        argEmail: String
+    ): Cocktail {
+        return Cocktail(
+            id,
+            cocktailDetailsData.name,
+            cocktailDetailsData.image,
+            cocktailDetailsData.alcoholic,
+            argFavorite,
+            argEmail
+        )
+    }
+
+    private fun setDetails(cocktailDetailsMap: Map<String, String>): String {
+        for (i in 1..15) {
+            val ingredientKey = "strIngredient$i"
+            val measureKey = "strMeasure$i"
+
+            val ingredient = cocktailDetailsMap[ingredientKey]
+            val measure = cocktailDetailsMap[measureKey]
+
+
+            if (!ingredient.isNullOrEmpty()) {
+                stringBuilderIngredient.append("$ingredient")
+                if (!measure.isNullOrEmpty()) {
```

Comment:
We need else part here as well, since 2 ingredients will be shown together in one line when we have no measure for the first one. Check cocktail GG.